src/cli/playbook/ipa.rs

```diff
@@ -4,37 +4,70 @@ use crate::{
     ff::Field,
     helpers::{query::QueryInput, transport::ByteArrStream},
     net::MpcHelperClient,
-    protocol::{attribution::AggregateCreditOutputRow, ipa::IPAInputRow, QueryId},
-    secret_sharing::IntoShares,
-    test_fixture::IPAInputTestRow,
+    protocol::{attribution::input::MCAggregateCreditOutputRow, ipa::IPAInputRow, QueryId},
+    secret_sharing::{
+        replicated::semi_honest::{AdditiveShare as Replicated, XorShare as XorReplicated},
+        IntoShares,
+    },
+    test_fixture::input::GenericReportTestInput,
 };
 use futures_util::future::try_join_all;
+use generic_array::{ArrayLength, GenericArray};
 use rand::{distributions::Standard, prelude::Distribution};
 use std::fmt::Debug;
+use std::ops::Add;
+use typenum::Unsigned;
 
 /// Semi-honest IPA protocol
 /// `(a, b)` will produce `a` * `b`.
 #[allow(clippy::missing_panics_doc)]
-pub async fn semi_honest<F, B>(
+pub async fn semi_honest<F, MK, BK>(
     input: InputSource,
     clients: &[MpcHelperClient; 3],
     query_id: QueryId,
 ) -> [Vec<impl Send + Debug>; 3]
 where
-    F: Field,
-    B: BitArray,
-    Standard: Distribution<F> + Distribution<B>,
+    F: Field + IntoShares<Replicated<F>>,
```

Comment:
wow this is rough. i need to redefine the trait bounds at the top level usage

Reply:
oh, is it worth doing then? we serialize it into a vec already, so there seem to be no point of trying to make it extremely efficient